From cfb4f880641df7066ba86ed9bfa29dc628854fb2 Mon Sep 17 00:00:00 2001
From: Dung Cao <dung@os.amperecomputing.com>
Date: Tue, 15 Nov 2022 20:48:50 +0700
Subject: [PATCH] pldmtool: fix pldmtool stuck in waiting forever response

Sometime one request may not receive the response. pldmtool should not
stuck in waiting forever the response. It should have a timeout
waiting for the response and retry to send the request before issuing
new pldm request.

Signed-off-by: Dung Cao <dung@os.amperecomputing.com>
Change-Id: I70b62b1ff30a8a715716db43867b479a5b214dfa
---
 src/requester/pldm.c | 35 ++++++++++++++++++++++++-----------
 1 file changed, 24 insertions(+), 11 deletions(-)

diff --git a/src/requester/pldm.c b/src/requester/pldm.c
index 932f55e..7c68486 100644
--- a/src/requester/pldm.c
+++ b/src/requester/pldm.c
@@ -6,9 +6,12 @@
 #include <sys/socket.h>
 #include <sys/types.h>
 #include <sys/un.h>
+#include <time.h>
 #include <unistd.h>
 
 const uint8_t MCTP_MSG_TYPE_PLDM = 1;
+const int NUMBER_OF_REQUEST_RETRIES = 2;
+const int RESPONSE_TIME_OUT = 1000;
 
 pldm_requester_rc_t pldm_open()
 {
@@ -148,18 +151,28 @@ pldm_requester_rc_t pldm_send_recv(mctp_eid_t eid, int mctp_fd,
 		return PLDM_REQUESTER_NOT_REQ_MSG;
 	}
 
-	pldm_requester_rc_t rc =
-	    pldm_send(eid, mctp_fd, pldm_req_msg, req_msg_len);
-	if (rc != PLDM_REQUESTER_SUCCESS) {
-		return rc;
-	}
-
-	while (1) {
-		rc = pldm_recv(eid, mctp_fd, hdr->instance_id, pldm_resp_msg,
-			       resp_msg_len);
-		if (rc == PLDM_REQUESTER_SUCCESS) {
-			break;
+	int i = 0;
+	pldm_requester_rc_t rc;
+	time_t t;
+	int sec;
+	while (i < (NUMBER_OF_REQUEST_RETRIES + 1)) {
+		rc = pldm_send(eid, mctp_fd, pldm_req_msg, req_msg_len);
+		if (rc != PLDM_REQUESTER_SUCCESS) {
+			return rc;
+		}
+		t = time(NULL);
+		while (1) {
+			rc = pldm_recv(eid, mctp_fd, hdr->instance_id,
+				       pldm_resp_msg, resp_msg_len);
+			if (rc == PLDM_REQUESTER_SUCCESS) {
+				return rc;
+			}
+			sec = (int)(time(NULL) - t);
+			if (sec >= (RESPONSE_TIME_OUT / 1000)) {
+				break;
+			}
 		}
+		i++;
 	}
 
 	return rc;
-- 
2.17.1

